---
title: "Class 10"
author: "Brian Wells (PID: A69026838)"
format: pdf
---

```{r}
string <- c("10", "100", 1, "1,000")
as.numeric(string)+1
```
>Q. Write a function to fix this non numeric table

```{r}
x <- string
as.numeric(gsub(",","", x))
```





```{r}
data.file <- "Data Export Summary.csv"
stats <- read.csv(data.file, row.names=1)
head(stats)
```

```{r}
rm.comma <- function(x) {
  as.numeric(gsub(",", "", x))
}

statsn <- apply(stats, 2, rm.comma)
head(statsn)
```
```{r}
rownames(statsn) <- rownames(stats)
statsn
```




of all of the protein sequences on pdb(251,600,768), only 183201 have been experimentally determined

```{r}
round(183201/251600768*100,2)
```

>Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.


```{r}
(sum(statsn[,1]) + sum(statsn[,2]))/sum(statsn[,7])*100
```
```{r}
totals <- apply(statsn, 2, sum)
round(totals/totals["Total"]*100,2)
```


>Q2: What proportion of structures in the PDB are protein?

```{r}
sum(statsn[1:3,7])/sum(statsn[,7])

```


>Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

I found 211,377 examples.


#pdb format


![](1HSG.png)



>Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

That is because the hydrogen is hidden, you can only see the oxygen. This is because the resolution is 2 angstrom, Hydrogen is only 1 A.



>Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have

Yes, HOH308


>Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.
Discussion Topic: Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?

![A lovely image](1HSGbetter.png)

#3. Introduction to Bio3D in R


```{r}
library(bio3d)
pdb <- read.pdb("1hsg")
```
```{r}
pdb
```



>Q7: How many amino acid residues are there in this pdb object?

198

>Q8: Name one of the two non-protein residues?

HOH


>Q9: How many protein chains are in this structure? 

2


```{r}
attributes(pdb)
```
```{r}
head(pdb$atom$resid)
```

#Predicting functional motions of a single structure

```{r}
adk <- read.pdb("6s36")
```
```{r}
adk
```
```{r}
# Perform flexiblity prediction
m <- nma(adk)
```
```{r}
plot(m)
```


```{r}
mktrj(m, file="adk_m7.pdb")
```






>Q10. Which of the packages above is found only on BioConductor and not CRAN? 

msa

>Q11. Which of the above packages is not found on BioConductor or CRAN?: 

bio3d-view

>Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?

TRUE


# Search and retrieve ADK structures


```{r}
library(bio3d)
aa <- get.seq("1ake_A")
```

```{r}
aa
```


>Q13. How many amino acids are in this sequence, i.e. how long is this sequence? 

214



# Search and retrieve ADK structures

```{r}
hits <- NULL
hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')

```

```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

#Align and superpose structures


```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```
```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
#plot(pdbs, labels=ids)

```

#Annotate collected PDB structures

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
```

```{r}
anno
```
#Principal component analysis


```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```



